src/span.rs

```diff
@@ -66,3 +70,254 @@ impl From<Range<usize>> for Span {
         }
     }
 }
+
+#[cfg(feature = "span")]
+extern crate smallvec;
+
+pub struct SpanContext(Span, Cow<'static, str>);
+
+impl SpanContext {
+    pub fn new<S>(span: Span, description: S) -> Self
+    where
+        Cow<'static, str>: From<S>,
+    {
+        Self(span, Cow::from(description))
+    }
+}
+
+#[derive(Debug)]
+pub struct WithSpan<E> {
+    inner: E,
+    #[cfg(feature = "span")]
+    spans: smallvec::SmallVec<[(Span, Cow<'static, str>); 8]>,
```

Comment:
Why are we using `SmallVec` here? It looks to be a wrapper around an error. If errors have heap allocations, that's totally cool. So here let's just use `Vec`.